src/D2L.CodeStyle.Analyzers/Pinning/PinnedAttributeAnalyzer.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Pinning {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class PinnedAttributeAnalyzer : DiagnosticAnalyzer {
+
+		private static readonly TypeKind[] TypeKindsSupported = new[] { TypeKind.Class, TypeKind.Interface };
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.PinnedTypesMustNotMove
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( OnCompilationStart );
+		}
+
+		private void OnCompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSymbolAction( AnalyzeSymbol,
+				SymbolKind.NamedType );
+		}
+
+		private static void AnalyzeSymbol( SymbolAnalysisContext context ) {
+			INamedTypeSymbol? pinnedAttributeSymbol = context.Compilation.GetTypeByMetadataName( PinnedAnalyzerHelper.PinnedAttributeName );

```

Comment:
nit: it's probably cached by Roslyn, but consider doing this in `OnCompilationStart` and passing it in here so it doesn't need to be called for every symbol.